heartbeat/scheduler/timerqueue/queue.go

```diff
@@ -0,0 +1,163 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package timerqueue
+
+import (
+	"container/heap"
+	"context"
+	"time"
+)
+
+// TimerTask represents a task run by the TimerQueue.
+type TimerTask struct {
+	fn    TimerTaskFn
+	runAt time.Time
+}
+
+// NewTimerTask creates a new TimerTask struct.
+func NewTimerTask(runAt time.Time, fn TimerTaskFn) *TimerTask {
+	return &TimerTask{runAt: runAt, fn: fn}
+}
+
+// TimerTaskFn is the function invoked by a TimerTask.
+type TimerTaskFn func(now time.Time)
+
+// TimerQueue represents a priority queue of timers.
+type TimerQueue struct {
+	th            *timerHeap
```

Comment:
hah, I was wondering where the timerHeap is used :)
How about namign the field `head` or `queue`?

Reply:
Unfortunately yes do to the way the heap package works, I think it rewrites the reference.

Reply:
See the examples in https://golang.org/pkg/container/heap/ for details.

Reply:
If you look at the IntHeap example, they indeed allocate. But looking at the PriorityQueue, the queue is not allocated. Instead of having a pointer here you can use `heap.Init(&queue)`. In the end it depends on how you implement the methods (pass by value vs by pointer) and how/when you pass the pointer to the heap functions.

Reply:
Interesting, thanks for the clarification, I'll make that change.